src/SDKs/Azure.Core/data-plane/Azure.Core/ReliableStream.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    public static class ReliableStream
+    {
+        public static async Task<Stream> Create(Func<long, Task<Response>> responseFactory, int maxRetries)
+        {
+            return new ReliableStreamImpl(await responseFactory(0), responseFactory, maxRetries);
+        }
+
+        private class ReliableStreamImpl : ReadOnlyStream
+        {
+            private readonly Func<long, Task<Response>> _responseFactory;
+
+            private readonly int _maxRetries;
+
+            private Stream _currentStream;
+
+            private long _position;
+
+            private int _retryCount;
+
+            private List<Exception> _exceptions;
+
+            public ReliableStreamImpl(Response initialResponse, Func<long, Task<Response>> responseFactory, int maxRetries)
+            {
+                _currentStream = initialResponse.ContentStream;
+                _responseFactory = responseFactory;
+                _maxRetries = maxRetries;
+                Length = _currentStream.Length;
+            }
+
+            public override long Seek(long offset, SeekOrigin origin)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                while (true)
+                {
+                    try
+                    {
+                        var result = await _currentStream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+                        _position += result;
+                        return result;
+                    }
+                    catch (Exception e)
+                    {
+                        await RetryAsync(e);
+                    }
+                }
+            }
+
+            private async Task RetryAsync(Exception exception)
+            {
+                if (_exceptions == null)
+                {
+                    _exceptions = new List<Exception>();
+                }
+
+                _exceptions.Add(exception);
+
+                // TODO: Verify exception type using ResponseClassifier and rethrow immediately for fatal
```

Comment:
You may want to consider doing this before potentially allocating the list, just in case you've got an unrecoverable exception like `OutOfMemory` or similar.